cmd/rook/ceph/operator.go

```diff
@@ -91,14 +99,176 @@ func startOperator(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		rook.TerminateFatal(err)
 	}
-
 	rookImage := rook.GetOperatorImage(context.Clientset, containerName)
 	serviceAccountName := rook.GetOperatorServiceAccount(context.Clientset)
 	op := operator.New(context, volumeAttachment, rookImage, serviceAccountName)
+	if isSecretPresent(context) {
+		initWebhook(context)
+	}
 	err = op.Run()
 	if err != nil {
 		rook.TerminateFatal(errors.Wrapf(err, "failed to run operator\n"))
 	}
 
 	return nil
 }
+
+func isSecretPresent(context *clusterd.Context) bool {
+	_, err := context.Clientset.CoreV1().Secrets("rook-ceph").Get("rook-admission-controller-secret", metav1.GetOptions{})
+	if err != nil {
+		logger.Infof("rook-admission-controller-secret not found. webhook initialization aborted")
+		return false
+	}
+	return true
+}
+
+func initWebhook(context *clusterd.Context) {
+
+	updateDeploymentWithVolumes(context)
+
+	if isMounted, err := isTLSVolumeMounted(context); isMounted && err == nil {
+		createWebhookService(context)
+		createValidatingWebhookConfig(context)
+		webhook.StartServer(context)
+	}
+
+}
+
+func isTLSVolumeMounted(context *clusterd.Context) (bool, error) {
+	rookDeploymentsClient := context.Clientset.AppsV1().Deployments("rook-ceph")
+
+	result, getErr := rookDeploymentsClient.Get("rook-ceph-operator", metav1.GetOptions{})
+
+	if getErr != nil {
+		logger.Errorf("failed to get latest version of deployment: %v", getErr)
+		return false, getErr
+	}
+
+	for _, v := range result.Spec.Template.Spec.Volumes {
+		if v.Name == "webhook-certs" {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func createWebhookService(context *clusterd.Context) {
+	serviceClient := context.Clientset.CoreV1().Services("rook-ceph")
+
+	webhookService := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rook-admission-controller-service",
+			Namespace: "rook-ceph",
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Port: 443,
+					TargetPort: intstr.IntOrString{
+						IntVal: 8443,
+					},
+				},
+			},
+			Selector: map[string]string{
+				"app": "rook-ceph-operator",
+			},
+		},
+	}
+	_, serErr := serviceClient.Create(&webhookService)
+	if serErr != nil && !kerrs.IsAlreadyExists(serErr) {
+		logger.Info("failed to create service")
+	}
+}
+
+func createValidatingWebhookConfig(context *clusterd.Context) {
+	configMap, mapErr := context.Clientset.CoreV1().ConfigMaps("kube-system").Get("extension-apiserver-authentication", metav1.GetOptions{})
```

Comment:
Remove hardcoded namespace.

Reply:
Do we have the RBAC permissions to create config in other namespace?

Reply:
Fixed
We are not creating this in other namespace. Its provided by the kubernetes cluster itself